### PR TITLE
Fix #9476: Support native TLS on Mac OS X

### PR DESCRIPTION
**Pull request by jacob-carlborg**
Second try at implementing native TLS on OS X. See https://github.com/D-Programming-Language/dmd/pull/5346 for reference.

Runtime part https://github.com/D-Programming-Language/druntime/pull/1523

### CODE REVIEW

**Comment by yebblies** [comment]
Looks like Martin's saying the druntime pull will break some existing shared library functionality.  Can that be fixed?

File: src/root/port.d [hunk 1]
```diff
@@ -121,6 +121,50 @@ extern (C++) struct Port
...
+     *
+     * Params:
+     *      str1 = the first string
+     *      str1Len = the length of the first string
+     *      str2 = the second string
+     *      str1Len = the length of the second string
+     *
+     * Returns: a new string which is the concatenation of the two strings
+     */
+    static char* concat(/*const*/ char* str1, uint str1Len, /*const*/ char* str2, uint str2Len)
```
Comment (yebblies) [inline src/root/port.d]: Actually, just get rid of concat and use `OutBuffer`.

Comment (jacob-carlborg) [inline src/root/port.d]: Ok.

**Comment by jacob-carlborg** [comment]
> Looks like Martin's saying the druntime pull will break some existing shared library functionality. Can that be fixed?

Yeah, I don't see how my change will break that. He'll have to give a better explanation.

File: src/root/port.d (hunk 1 above)
Comment (jacob-carlborg) [inline src/root/port.d]: Updated to use `OutBuffer`.

**Comment by jacob-carlborg** [comment]
Is the FreeBSD test expected to fail sometimes? The test output contains `test fails depending on environment`.

**Comment by jacob-carlborg** [comment]
I think we cleared up the misunderstanding with the shared library support that Martin was concerned about https://github.com/D-Programming-Language/druntime/pull/1461#discussion_r57446927.

**Comment by jacob-carlborg** [comment]
@yebblies @MartinNowak could we try this again? This time all test are passing except for 64bit OS X, which is expected since the druntime part is required as well.

**Comment by jacob-carlborg** [comment]
@yebblies @MartinNowak @WalterBright @9rnsr please review.

**Comment by smolt** [comment]
@jacob-carlborg - something to be aware of.  I ran into a possible bug related to OS X TLS after updating to latest Xcode 7.3.  See ldc-developers/ldc#1444.  Error is `ld: section __DATA/__thread_bss extends beyond end of file`. It might be worth waiting a bit until the problem is understood because dmd is immune right now with its own TLS implementation.

@deadalnix - you filed an llvm bug on this I think too.  Do you know more on it?  My hunch is a linker bug, but I am not sure.

**Comment by jacob-carlborg** [comment]
Hmm, that's unfortunate.

**Comment by deadalnix** [comment]
@smolt Yes I'm not sure what the problem is. It may well be a binutil problem rather than an llvm one. I'm just not sure.

**Comment by smolt** [comment]
I was informed my Apple radar bug 25769807 is duplicate of 24667656 which is Closed. I assume that means there will be a fix in an upcoming Xcode release.

**Comment by deadalnix** [comment]
So this is an XCode bug ?

**Comment by smolt** [comment]
Yes, in that Xcode includes entire clang/llvm compile chain.  I don't know what specific component has bug, they don't tell me.

**Comment by deadalnix** [comment]
I bet on the linker.

**Comment by deadalnix** [comment]
@yebblies Were you able to solve the linker problem ? Was something needed or just updating the linker did the trick ?

**Comment by yebblies** [comment]
Nope.

**Comment by smolt** [comment]
I got an update to Xcode 7.3.1.  I'll have to see if the error goes away when I rerun test suite against LDC.

**Comment by deadalnix** [comment]
OK software update somehow screwed up the update on my mac. I was able to get it back in shape and the bug is gone as well. Thanks everybody.

**Comment by bitwise-github** [comment]
Has anyone benchmarked to check relative performance of accessing TLS vars?

**Comment by smolt** [comment]
Also confirmed bug is gone Xcode 7.3.1.

**Comment by MartinNowak** [comment]
WAT? Please prepare a changelog entry mentioning the minimum required OSX and XCode versions, and add some micro-benchmark numbers along with it.

**Comment by deadalnix** [comment]
Don't freak out. OSX broke its linker and the fixed it. It doesn't require
a cutting edge version of osx/xcode.

**Comment by MartinNowak** [comment]
Whatever version it is, you have to tell this to people not involved in this PR.
How would they ever find out other than running through the same trouble as you?

**Comment by smolt** [comment]
I believe the bug appeared with linker delivered in Xcode 7.3 and then fixed with Xcode 7.3.1.  It affected C code too.  @jacob-carlborg I think just a warning to avoid from Xcode 7.3 in change log.  We should do that in LDC release notes too.

**Comment by deadalnix** [comment]
By linking any program that has tls greater than some non trivial size, in
D or otherwize.

OSX even has an auto update for the thing (it just somehow failed on my
machine).

**Comment by MartinNowak** [comment]
Well just add the changelog entry, you should inform people about using native TLS anyhow.

**Comment by jacob-carlborg** [comment]
@MartinNowak https://github.com/dlang/dmd/pull/5740

**Comment by jacob-carlborg** [comment]
@MartinNowak this updates the requirements for OS X as well: https://github.com/dlang/dlang.org/pull/1292.

**Pull request merged**